lib/options.c

```diff
@@ -40,371 +40,249 @@
            != -1)
     {
         switch (c) {
-        case 1: { /* positional arguments */
-            char **tmp = append_arg_to_vector (&new_argc, new_argv, optarg);
-            if (tmp == NULL) {
-                free(new_argv);
-                return 2;
-            }
-            new_argv = tmp;
-        }
-            break;
         case 'T':
-            common_opts->tcti_type = tcti_type_from_name (optarg);
-            break;
-#ifdef HAVE_TCTI_DEV
-        case 'd':
-            common_opts->device_file = optarg;
-            break;
-#endif
-#ifdef HAVE_TCTI_SOCK
-        case 'R':
-            common_opts->socket_address = optarg;
+            tcti_name = optarg;
             break;
-        case 'p': {
-            bool res = tpm2_util_string_to_uint16(optarg, &common_opts->socket_port);
-            if (!res) {
-                LOG_ERR("Could not convert port to a 16 bit unsigned number, "
-                        "got: %s", optarg);
-                free(new_argv);
-                return 2;
-            }
-        }   break;
-#endif
-#ifdef HAVE_TCTI_TABRMD
-        /* No options for this TCTI yet. */
-#endif
         case 'h':
-            common_opts->help = true;
+            // TODO
+            //execute_man();
             break;
         case 'V':
-            common_opts->verbose = true;
+            flags->verbose = 1;
             break;
         case 'Q':
-            common_opts->quiet = true;
+            flags->quiet = 1;
             break;
         case 'v':
-            common_opts->version = true;
+            // TODO
+            //showversion();
             break;
-        case '?': {
-            char **tmp = append_arg_to_vector (&new_argc, new_argv, argv[optind - 1]);
-            if (tmp == NULL) {
-                free(new_argv);
-                return 2;
+        case ':':
+            LOG_ERR("Argument %c needs a value!", optopt);
+            goto out;
+        case '?':
+            LOG_ERR("Unknown Argument: %c", optopt);
+            result = false;
+            goto out;
+        default:
+            result = tool_opts->on_opt(c, optarg);
+            if (!result) {
+                goto out;
             }
-            new_argv = tmp;
         }
-            break;
+    }
+
+    char **tool_args = &argv[optind];
+    int tool_argc = argc - optind;
+
+    /* have args and a handler to process */
+    if (tool_argc && tool_opts->on_arg) {
+        result = tool_opts->on_arg(tool_argc, tool_args);
+    /* have args and no handler, error condition */
+    } else if (tool_argc && !tool_opts->on_arg) {
+        result = false;
+        goto out;
+    }
+
+    size_t i;
+    bool found = false;
+    for(i=0; i < ARRAY_LEN(tcti_map_table); i++) {
+
+        char *name = tcti_map_table[i].name;
+        tcti_init init = tcti_map_table[i].init;
+        if (!strcmp(tcti_name, name)) {
+            found = true;
+            *tcti = init(tcti_opts);
```

Comment:
You haven't implemented --tcti-args right? That's what should be in tcti_opts

Reply:
yes